tron/kubernetes.py

```diff
@@ -148,35 +152,71 @@ def handle_event(self, event: Event) -> None:
+                        # https://github.com/kubernetes/kubernetes/issues/100047#issuecomment-797624208) and then decide that
+                        # these Pods should be phase=Succeeded with an exit code of 0 - even though the container never actually
+                        # started. So far, we've noticed that when this happens, the finished_at and reason fields will be None
+                        # and thus we'll check for at least one of these conditions to detect an abnormal exit and actually "fail"
+                        # the affected action
+                        # NOTE: hopefully this won't change too drastically in future k8s upgrades without the actual problem (incorrect
+                        # success) being fixed :p
+                        if state_termination_metadata.get("exitCode") == 0 and (
+                            state_termination_metadata.get("finishedAt") is None
+                            and state_termination_metadata.get("reason") is None
+                        ):
+                            exit_code = exitcode.EXIT_KUBERNETES_ABNORMAL
+                            self.log.warning("Container never started due to a Kubernetes/infra flake!")
+                            self.log.warning(
+                                f"If automatic retries are not enabled, run `tronctl retry {self.id}` to retry."
+                            )
+                    elif k8s_type in KUBERNETES_FAILED_TYPE:
+                        # Handling spot terminations
+                        if (
+                            last_state_termination_metadata.get("exitCode") == 137
+                            and last_state_termination_metadata.get("reason") == "ContainerStatusUnknown"
+                        ):
+                            exit_code = exitcode.EXIT_KUBERNETES_SPOT_INTERRUPTION
+                            self.log.warning("Tronjob failed due to spot interruption.")
+                        # Handling K8s scaling down a node
+                        elif state_termination_metadata.get("exitCode") == 143 and (
+                            state_termination_metadata.get("reason") == "Error"
+                        ):
+                            exit_code = exitcode.EXIT_KUBERNETES_NODE_SCALEDOWN
+                            self.log.warning("Tronjob failed due to Kubernetes scaling down a node.")
+                        if exit_code in KUBERNETES_EXIT_CODE_EXCEPTIONS:
+                            self.log.warning(
+                                f"If automatic retries are not enabled, run `tronctl retry {self.id}` to retry."
+                            )
+                            self.log.warning(
+                                "If this action is idempotent, then please consider enabling automatic retries for your action. If your action is not idempotent, then please configure this action to run on the stable pool rather than the default."
+                            )
+                    else:
```

Comment:
i'm not sure we'll actually hit this case: the `if` / `elif` cases above this capture all the possible cases, right?

I think we'd want this at a slight different level: if the flow of the existing code was something like:
```
if abnormal_exit:
    ...
elif successful_run:
    ...
elif spot_interruption or node_scaledown:
    ...
then we'd want something like 
```
if abnormal_exit:
    ...
elif successful_run:
    ...
elif spot_interruption or node_scaledown:
    ...
else: # this case, where there was a "normal" failure
    ...
```

right now the logic is something like the following, though:
```
if success:
    ...
elif failure:
    ...
else: # block we're talking about
    ...
```
so it doesn't seem like we'll ever actually hit this case 

to verify: i think we could setup a tronjob with a tiny memory limit and then have it execute a python command that just does something like `print(bytearray(512000000))` - this should result in an oomkill and tron showing 137 as an exit code if i'm reading the code incorrectly - but if i'm reading it correctly, this test case will never hit this code

Reply:
you're right, I had this in the wrong place